### PR TITLE
[1020] missing details component

### PR DESCRIPTION
**Pull request by defong**
### Context

Adding logic to the `check-details` page which renders a link and text to sections that haven't been started, or are uncompleted.

### Changes proposed in this pull request

- Add new missing details component
- Add custom CSS
- Create logic to make component dynamic
- Add missing details text to locale
- Create view preview
- Add styling for error messages

### Guidance to review
- Navigate to the check details page: https://rtt-review-pr-540.herokuapp.com/trainees/nXZr9qmsuEQ6BDPEm2qZjPpU/check-details
- Click the `Start section` link and complete the section without ticking the 'completed' checkbox at the end
- The text and link should have changed on the `/check-details` page
- Repeat for each section
- Check the view previews
- Check styling for error messages

#### Dumb components
`IncompleteSection::View`
- only get feed with values
- it a wrapper for `GovukComponent::InsetText`
- and displays as seen in the links below, there is a default and an error
https://rtt-review-pr-540.herokuapp.com/view_components/incomplete_section/view/default
https://rtt-review-pr-540.herokuapp.com/view_components/incomplete_section/view/error

#### Smart components
`Trainees::Sections::View`
- it decides which component to renders
- its a wrapper for any confirmation components as well as `IncompleteSection::View` over various states
- it deals with trainee progress
https://rtt-review-pr-540.herokuapp.com/view_components/trainees/sections/view

### Screenshots
image
image
image
image
image

### CODE REVIEW

**Comment by avinhurry** [comment]
[[M] Give the review record page component state (or another component)](https://trello.com/c/8bTmKcGX/1020-m-give-the-review-record-page-component-state-or-another-component)

File: config/locales/en.yml [hunk 1]
```diff
@@ -147,6 +147,50 @@ en:
         withdrawn: Trainee withdrawn
         deferred: Trainee deferred
         qts_awarded: Trainee awarded QTS
+        programme_details: Programme details not started
+    sections:
+      personal_details:
+        not_started: 
+          title: Personal details not started
+          link_text: Start section
```
Comment (AbigailMcP) [inline config/locales/en.yml:155]: Could we have just one translation for `link_text.not_stated` and `link_text.in_progress`, seeing as they're always either 'Start section' or 'Continue section'?

Seems like a lot of repetition if we're not expecting them to individually change.

File: spec/factories/trainees.rb [hunk 2]
```diff
@@ -44,6 +44,51 @@
...
+      commencement_date { nil }
+    end
+
+    trait :in_progress do
+      with_programme_details
+      with_start_date
+      degrees { [build(:degree, :uk_degree_with_details)] }
+    end
+
+    trait :completed do
```
Comment (AbigailMcP) [inline spec/factories/trainees.rb:77]: Ah I think I also added a similar trait `ready_for_trn_submission` - but yours are neater. Please keep your definitions when you do the final merge.

File: app/views/trainees/check_details/show.html.erb [hunk 3]
```diff
@@ -42,16 +42,21 @@
       Personal details and education
     </h2>
 
-    <%= render Trainees::Confirmation::PersonalDetails::View.new(trainee: @trainee) %>
-    <%= render Trainees::Confirmation::ContactDetails::View.new(trainee: @trainee) %>
-    <%= render Trainees::Confirmation::Diversity::View.new(trainee: @trainee) %>
-    <%= render Trainees::Confirmation::Degrees::View.new(trainee: @trainee, show_add_another_degree_button: false, show_delete_button: true) %>
+    <%= render Trainees::Sections::View.new(trainee: @trainee, trn_submission: @trn_submission, section: :personal_details) %>
```
Comment (AbigailMcP) [inline app/views/trainees/check_details/show.html.erb:45]: I wonder whether passing the whole `@trn_submission` form object is a little overkill here, when we could pass in the error boolean and the section's state?

File: app/components/incomplete_section/view.rb [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+  attr_accessor :title, :url, :link_text, :error
+
+  def initialize(title:, link_text:, url:, error: false)
+    @url = url
+    @link_text = link_text
+    @title = title
+    @error = error
+  end
+
+  def text
```
Comment (AbigailMcP) [inline app/components/incomplete_section/view.rb:13]: This method isn't used.

**Review by gpeng — CHANGES_REQUESTED** [review]
Just a couple of little things. I think we should merge this for now but we can definitely simplify by making the details component more general and moving all of the data logic out. We could also refactor to use a view object to present the data and then the 'smart' component test wouldn't have to deal with all the states. It would just test that it renders the data it's passed. Also thinking that this will have to be updated to deal with different routes...

File: app/views/trainees/check_details/show.html.erb (hunk 3 above)
Comment (avinhurry) [reply app/views/trainees/check_details/show.html.erb:45]: Had a discussion offline and decided to leave this since we are planning to do another refactor

**Comment by ndgallon73** [comment]
Hi, I found an error when following the above instructions to partially complete all sections bar the tickbox. Request A TRN get  an error saying all section must be completed. I then went in to the personal details section to tick the box and got the attached e
Screenshot 2021-02-19 at 13 46 48
rror?

**Comment by edwardhorsford** [comment]
Looking good - but I notice something we've missed. We have multiple links with the same link text, which we should avoid.

We can fix this by adding hidden text as part of each link - eg `Complete section <hidden> personal details</hidden>` - this would be similar to what the `change` links do in the summary lists.

**Comment by avinhurry** [comment]
> Hi, I found an error when following the above instructions to partially complete all sections bar the tickbox. Request A TRN get an error saying all section must be completed. I then went in to the personal details section to tick the box and got the attached e
> Screenshot 2021-02-19 at 13 46 48
> rror?

I think @darokel has fixed this as part of a separate, when we rebase with master this will be fixed

**Comment by avinhurry** [comment]
> Looking good - but I notice something we've missed. We have multiple links with the same link text, which we should avoid.
> 
> We can fix this by adding hidden text as part of each link - eg `Complete section <hidden> personal details</hidden>` - this would be similar to what the `change` links do in the summary lists.

Thanks @edwardhorsford. Changes made

**Pull request merged**